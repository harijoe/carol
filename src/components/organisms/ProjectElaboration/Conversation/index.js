import React, { Component, PropTypes } from 'react'
import styled from 'styled-components'

import Question from './Question'
import Response from './Response'
import QuickReplies from './QuickReplies'

const Wrapper = styled.div`
  maxHeight: 400px;
  overflow: scroll;
  bottom: 0;
`

class Conversation extends Component {
  static propTypes = {
    conversation: PropTypes.arrayOf(
      PropTypes.shape({
        message: PropTypes.shape({
          text: PropTypes.string.isRequired,
          quick_replies: PropTypes.array,
        }).isRequired,
        response: PropTypes.string,
      }),
    ),
    reply: PropTypes.func,
  }

  componentDidMount() {
    this.scrollTop()
  }

  componentDidUpdate() {
    this.scrollTop()
  }

  scrollTop() {
    this.history.scrollTop = this.history.scrollHeight
  }

  render() {
    const { conversation, reply } = this.props

    return (
      <Wrapper innerRef={(ref) => { this.history = ref }}>
        {
          conversation.map((field, index) => (
            <div key={index}>
              <Question question={field.message.text} />
              {
                conversation.length - 1 === index ?
                  <QuickReplies
                    quick_replies={field.message.quick_replies}
                    reply={reply}
                    response={field.response}
                  />
                  : null
              }
              <Response response={field.response} />
            </div>
          ))
        }
      </Wrapper>
    )
  }
}

export default Conversation
