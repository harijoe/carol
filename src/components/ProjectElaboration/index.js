import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import isFirstChatbotStep from 'utils/isFirstChatbotStep'
import { breakpoint } from 'utils/style'

import Grid from 'components/Grid'
import Conversation from './containers/Conversation'
import Form from './organisms/Form'

const StyledGrid = styled(Grid)`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  height: 100vh;
  max-height: 100vh;
  padding-bottom: 0;
  padding-top: 5.6rem;
  background: white;

  ${breakpoint('l')`
    max-width: none;
  `};
`

const ProjectElaboration = ({ activeConversation, reply }) => {
  const quickReplies = activeConversation.length > 0 ? activeConversation[activeConversation.length - 1].message.quick_replies : null
  const enableBack = !isFirstChatbotStep(activeConversation)

  return (
    <StyledGrid narrow>
      <div>
        <Conversation {...{ activeConversation, reply }} />
        <Form reply={reply} disabled={quickReplies != null ? quickReplies.length !== 0 : true} enableBack={enableBack} />
      </div>
    </StyledGrid>
  )
}

ProjectElaboration.propTypes = {
  activeConversation: PropTypes.arrayOf(
    PropTypes.shape({
      message: PropTypes.shape({
        quick_replies: PropTypes.array,
      }),
      answer: PropTypes.shape({
        text: PropTypes.string,
      }),
    }),
  ),
  reply: PropTypes.func,
}

export default ProjectElaboration
