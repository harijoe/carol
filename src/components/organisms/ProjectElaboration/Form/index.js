import React, { Component, PropTypes } from 'react'
import { Field, reduxForm } from 'redux-form'
import styled from 'styled-components'
import { FormattedMessage, injectIntl, intlShape } from 'react-intl'

import messages from 'utils/messages'
import { RenderField, Button } from 'components'

const StyledForm = styled.form`
  box-sizing: border-box;
  padding: 1rem;
`

class Form extends Component {
  componentDidUpdate() {
    this.field.focus()
  }

  render() {
    const {
      reply,
      intl: { formatMessage },
      handleSubmit,
      reset,
      pristine,
      submitting,
      disabled,
    } = this.props
    const submit = (values) => {
      reply(values.response)
      reset()
    }

    return (
      <StyledForm onSubmit={handleSubmit(submit)}>
        <Field
          disabled={disabled}
          name="response"
          component={RenderField}
          placeholder={formatMessage(messages('project.elaboration.response').label)}
          autoFocus
          innerRef={(field) => { this.field = field }}
        />
        <Button disabled={pristine || submitting} type="submit">
          <FormattedMessage id="project.elaboration.submit" tagName="span" />
        </Button>
      </StyledForm>
    )
  }
}

Form.propTypes = {
  reply: PropTypes.func.isRequired,
  intl: intlShape,
  handleSubmit: PropTypes.func,
  reset: PropTypes.func,
  pristine: PropTypes.bool,
  submitting: PropTypes.bool,
  disabled: PropTypes.bool,
}

export default reduxForm({ form: 'ProjectElaboration' })(injectIntl(Form))
