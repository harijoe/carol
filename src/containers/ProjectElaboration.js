import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { push } from 'react-router-redux'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import {
  projectElaborationReply,
  projectElaborationConversationCurrent,
  projectElaborationConversationsSelect,
} from 'store/actions'
import { fromProjectElaboration, fromContext } from 'store/selectors'

import { ProjectElaboration } from 'components'


class ProjectElaborationContainer extends Component {
  static propTypes = {
    request: PropTypes.func,
  }

  componentWillMount() {
    this.props.request()
  }

  render() {
    return (
      <ProjectElaboration
        {...this.props}
      />
    )
  }
}

const mapStateToProps = state => ({
  activeConversation: fromProjectElaboration.getConversation(state),
  hero: fromProjectElaboration.getHero(state),
  conversations: fromProjectElaboration.getConversations(state),
  hasConversations: fromProjectElaboration.hasConversations(state),
  locale: fromContext.getLocale(state),
})

const mapDispatchToProps = dispatch => bindActionCreators({
  reply: (text, payload) => projectElaborationReply.request(text, payload),
  request: () => projectElaborationConversationCurrent.request(),
  selectConversation: authType => projectElaborationConversationsSelect.request(authType),
  redirectTo: url => push(url),
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ProjectElaborationContainer)
