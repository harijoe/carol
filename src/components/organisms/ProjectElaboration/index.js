import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { Grid } from 'components'
import Conversation from './organisms/Conversation'
import Form from './organisms/Form'

const StyledGrid = styled(Grid)`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  height: 100vh;
  max-height: 100vh;
  padding-bottom: 0;
  padding-top: 5.6rem;
`


const ProjectElaboration = ({ conversation, reply }) => {
  const quickReplies = conversation.length > 0 ? conversation[conversation.length - 1].message.quick_replies : null

  return (
    <StyledGrid narrow>
      <Conversation conversation={conversation} reply={reply} />
      <Form reply={reply} disabled={quickReplies != null ? quickReplies.length !== 0 : true} />
    </StyledGrid>
  )
}

ProjectElaboration.propTypes = {
  conversation: PropTypes.arrayOf(
    PropTypes.shape({
      message: PropTypes.shape({
        quick_replies: PropTypes.array,
      }),
    }),
  ),
  reply: PropTypes.func,
}

export default ProjectElaboration
