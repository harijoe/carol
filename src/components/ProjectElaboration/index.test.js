import React from 'react'
import { shallow } from 'enzyme'

import mockIntl from 'mocks/intlMock'
import ProjectElaboration from './'

const wrap = (props = {}) => shallow(mockIntl(<ProjectElaboration {...props} />))

it('renders ProjectElaboration', () => {
  expect(
    wrap({
      activeConversation: [
        {
          message: {
            text: 'message',
            quick_replies: [],
          },
        },
      ],
    }),
  ).toMatchSnapshot()
})
