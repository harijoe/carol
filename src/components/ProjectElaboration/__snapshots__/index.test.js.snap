// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProjectElaboration 1`] = `
<ProjectElaboration
  activeConversation={
    Array [
      Object {
        "message": Object {
          "quick_replies": Array [],
          "text": "message",
        },
      },
    ]
  }
/>
`;
